util/build_configs/setenv-example-3.bash

```diff
@@ -0,0 +1,247 @@
+
+    export CHPL_HOST_PLATFORM=cray-xc
+    export CHPL_TARGET_PLATFORM=cray-xc
+##  export CHPL_REGEXP=re2      # re2 required for mason
+    export CHPL_LOCAL_MODEL=flat
+    export CHPL_COMM=none
+##  export CHPL_COMM_SUBSTRATE=none
+##  export CHPL_TASKS=qthreads
+    export CHPL_LAUNCHER=none
+##  export CHPL_LLVM=none       # llvm requires py27 and cmake
+##  export CHPL_AUX_FILESYS=none
+
+
+    # Show the initial/default Chapel build config with printchplenv
+
+    if [ -n "$verbose" ]; then
+        log_debug "Chapel printchplenv, with initial env:"
+        $CHPL_HOME/util/printchplenv --all --no-tidy || echo >&2 ignore error
+    fi
+
+    # NOTE: The --target-compiler values used in this setenv project are never seen by Chapel make.
+    # Instead, they drive module load commands in the setenv callback
+
+
+    # Chapel compiler
+
+    log_info "Start build_configs $dry_run $verbose compiler"
+
+    $cwd/build_configs.py -p $dry_run $verbose -s $cwd/$setenv -l "$project.compiler.log" \
+        --target-compiler=compiler compiler
+
+
+    # Chapel runtimes
+
+    compilers="gnu,intel"
+    comms="none,ugni"
+    launchers="none,slurm-srun"
+
+    log_info "Start build_configs $dry_run $verbose # no make target"
+
+    $cwd/build_configs.py -p $dry_run $verbose -s $cwd/$setenv -l "$project.runtime.log" \
+        --target-compiler=$compilers --comm=$comms --launcher=$launchers
+
+
+    # Chapel tools (ie, mason, chpldoc, test-venv)
+
+##  # Building Chapel tools requires a working internet connection and either:
+##  # - modern version of libssl that supports TLS 1.1
+##  # - local workarounds such as the variables shown in the "tools" callback, below.
+##  #
+##  # Commented out here to make this setenv example more portable.
+##
+##  tools_targets="test-venv chpldoc mason"
+##
+##  log_info "Start build_configs $dry_run $verbose $tools_targets"
+##
+##  $cwd/build_configs.py $dry_run $verbose -s $cwd/$setenv -l "$project.tools.log" \
+##      --target-compiler=tools $tools_targets
+
+
+    # make clean
+
+    clean_targets="cleanall"
+
+    log_info "Start build_configs $dry_run $verbose $clean_targets"
+
+    $cwd/build_configs.py $dry_run $verbose -s $cwd/$setenv -l "$project.clean.log" \
+        --target-compiler=compiler $clean_targets
+
+    log_info "End $setenv"
+else
+
+# ===============
+# setenv callback
+# ===============
+
+    setenv=$( basename "${BASH_SOURCE[0]}" )
+    cwd=$( cd $(dirname "${BASH_SOURCE[0]}") && pwd )
+    source $cwd/functions.bash
+
+    log_info "Begin callback $setenv"
+    log_debug "with config=$BUILD_CONFIGS_CALLBACK"
+
+# =========================
+# setenv callback functions
+# =========================
+
+    source $cwd/module-functions.bash
+
+    # check module system
+    ck_module_list
+
+    if [ -n "$verbose" ] ; then
+        log_debug "Module list (sorted), after build_configs, before setenv:"
+        module list -t 2>&1 | tail -n +2 | sort
```

Comment:
Not sure how much it matters, but if we wanted to isolate all calls to `module` within `module-functions.sh`, we might make this a function, e.g. `sorted_modules`.

Reply:
That makes sense. Didn't think of it.